sdks/java/io/amqp/src/main/java/org/apache/beam/sdk/io/amqp/AmqpIO.java

```diff
@@ -0,0 +1,471 @@
+
+  }
+
+  /**
+   * A {@link PTransform} to send messages using AMQP 1.0 protocol.
+   */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<String>, PDone> {
+
+    @Nullable abstract List<String> addresses();
+    @Nullable abstract String subject();
+
+    abstract Builder builder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setAddresses(List<String> addresses);
+      abstract Builder setSubject(String subject);
+      abstract Write build();
+    }
+
+    /**
+     * Define the AMQP addresses where to send messages.
+     */
+    public Write withAddresses(List<String> addresses) {
+      checkArgument(addresses != null, "AmqpIO.write().withAddresses(addresses) called with "
+          + "null addresses");
+      checkArgument(!addresses.isEmpty(), "AmpqIO.write().withAddresses(addresses) called with "
+          + "empty addresses list");
+      return builder().setAddresses(addresses).build();
+    }
+
+    /**
+     * Define the AMQP messages subject.
+     */
+    public Write withSubject(String subject) {
+      checkArgument(subject != null, "AmqpIO.write().withSubject(subject) called with null "
+          + "subject");
+      return builder().setSubject(subject).build();
+    }
+
+    @Override
+    public PDone expand(PCollection<String> input) {
+      input.apply(ParDo.of(new WriteFn(this)));
+      return PDone.in(input.getPipeline());
+    }
+
+    @Override
+    public void validate(PipelineOptions pipelineOptions) {
+      checkState(addresses() != null, "AmqIO.write() requires addresses list to be set via "
+          + "withAddresses(addresses)");
+      checkState(!addresses().isEmpty(), "AmqIO.write() requires a non-empty addresses list to be"
+          + " set via withAddresses(addresses)");
+    }
+
+    @Override
+    public void populateDisplayData(DisplayData.Builder builder) {
+      StringBuilder stringBuilder = new StringBuilder();
+      for (String address : addresses()) {
+        stringBuilder.append(address).append(" ");
+      }
+      builder.add(DisplayData.item("addresses", stringBuilder.toString()));
+      builder.addIfNotNull(DisplayData.item("subject", subject()));
+    }
+
+    private static class WriteFn extends DoFn<String, Void> {
+
+      private final Write spec;
+
+      private transient Messenger messenger;
+
+      public WriteFn(Write spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void setup() throws Exception {
+        messenger = Messenger.Factory.create();
+        messenger.start();
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext processContext) throws Exception {
+        String element = processContext.element();
+        for (String address : spec.addresses()) {
+          Message message = Message.Factory.create();
+          message.setAddress(address);
+          if (spec.subject() != null) {
+            message.setSubject(spec.subject());
```

Comment:
Seems a bit weird that this is the only field we allow to set except body. Would be less weird if the IO took Message's rather than String :)